### ISSUE TITLE
Erroneous MissingOutputException / UnknownPod errors using Kubernetes executor on GKE

### ISSUE THREAD

**Issue by cademirch** [snakemake/snakemake#1405]
**Snakemake version**

6.15.5
**Describe the bug**

Executing `snakemake` with `--kubernetes` on GCP I am running into erroneous `MissingOutputException`. Looking at the logs it seems that the jobs ended successfully and the output files were successfully uploaded to the bucket. The reported incomplete files appear to be intact and look as expected.
**Logs**

```
Traceback (most recent call last):
  File "/home/cade/mambaforge/envs/snakemake/lib/python3.10/site-packages/snakemake/dag.py", line 493, in check_and_touch_output
    wait_for_files(
  File "/home/cade/mambaforge/envs/snakemake/lib/python3.10/site-packages/snakemake/io.py", line 824, in wait_for_files
    raise IOError(
OSError: Missing files after 5 seconds:
ccgp-workflow-results/results/Artemisiospiza/GCA_021963965.1/01_mappedReads/MVZCCGP-AbeI_A01_final.bam
ccgp-workflow-results/results/Artemisiospiza/GCA_021963965.1/01_mappedReads/MVZCCGP-AbeI_A01_final.bam.bai
ccgp-workflow-results/benchmarks/Artemisiospiza/GCA_021963965.1/dedup/MVZCCGP-AbeI_A01.txt

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/cade/mambaforge/envs/snakemake/lib/python3.10/site-packages/snakemake/scheduler.py", line 529, in _finish_jobs
    self.get_executor(job).handle_job_success(job)
  File "/home/cade/mambaforge/envs/snakemake/lib/python3.10/site-packages/snakemake/executors/__init__.py", line 876, in handle_job_success
    super().handle_job_success(
  File "/home/cade/mambaforge/envs/snakemake/lib/python3.10/site-packages/snakemake/executors/__init__.py", line 265, in handle_job_success
    job.postprocess(
  File "/home/cade/mambaforge/envs/snakemake/lib/python3.10/site-packages/snakemake/jobs.py", line 1027, in postprocess
    self.dag.check_and_touch_output(
  File "/home/cade/mambaforge/envs/snakemake/lib/python3.10/site-packages/snakemake/dag.py", line 500, in check_and_touch_output
    raise MissingOutputException(
snakemake.exceptions.MissingOutputException: Job Missing files after 5 seconds:
ccgp-workflow-results/results/Artemisiospiza/GCA_021963965.1/01_mappedReads/MVZCCGP-AbeI_A01_final.bam
ccgp-workflow-results/results/Artemisiospiza/GCA_021963965.1/01_mappedReads/MVZCCGP-AbeI_A01_final.bam.bai
ccgp-workflow-results/benchmarks/Artemisiospiza/GCA_021963965.1/dedup/MVZCCGP-AbeI_A01.txt
This might be due to filesystem latency. If that is the case, consider to increase the wait time with --latency-wait.
Job id: 9 completed successfully, but some output files are missing. 9

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/cade/mambaforge/envs/snakemake/lib/python3.10/site-packages/snakemake/__init__.py", line 701, in snakemake
    success = workflow.execute(
  File "/home/cade/mambaforge/envs/snakemake/lib/python3.10/site-packages/snakemake/workflow.py", line 1077, in execute
    success = self.scheduler.schedule()
  File "/home/cade/mambaforge/envs/snakemake/lib/python3.10/site-packages/snakemake/scheduler.py", line 440, in schedule
    self._finish_jobs()
  File "/home/cade/mambaforge/envs/snakemake/lib/python3.10/site-packages/snakemake/scheduler.py", line 534, in _finish_jobs
    self._handle_error(job)
  File "/home/cade/mambaforge/envs/snakemake/lib/python3.10/site-packages/snakemake/scheduler.py", line 615, in _handle_error
    self.running.remove(job)
KeyError: dedup
```
**Minimal example**

I have not been able to reliably recreate this issue. 
**Additional context**

Would appreciate any insight on how Snakemake determines what files may be missing as that seems to be the best place to start. Digging through the source code myself I could not quite figure it out. I have also tried increasing `--latency-wait ` to 900 with no avail.

**Comment by cademirch** [comment]
Might be related to #1378

**Comment by cademirch** [comment]
This has become quite problematic for running large workflows on GKE. Including `--restart-times 3` leads to Snakemake retrying the job it believes failed (despite the output existing in the bucket) over and over again until it has tried  and 'failed' 3 times, until the workflow exits. Curious if any others have experienced this.

**Comment by cademirch** [comment]
Did some more digging into this. Comparing the logs of a successful pod vs. a 'failed' unknown pod it seems that the issue may be network related? Though I am not entirely sure. I am pasting below the relevant logs from a good pod and a bad pod below.

To me, it seems like the API call to delete the pod isn't making to to the 'bad' pod and it gets killed by the cluster. Where as in the 'good' pod I can see where the API call is received from my IP address (starred out). Until I can figure this out the `--kubernetes` option is pretty much unusable. I am curious if others have seen this issue running large workflows with GKE? Would also appreciate any insight on this. I've also attached the full logs for both pods (they're .txt files because github doesn't allow json uploads).

[bad_finished_pod.txt](https://github.com/snakemake/snakemake/files/8149426/bad_finished_pod.txt)
[safe_finished_pod.txt](https://github.com/snakemake/snakemake/files/8149427/safe_finished_pod.txt)


###'Failed'/Unknown pod:
```
{
    "textPayload": "1 of 1 steps (100%) done\n",
    "insertId": "4upxcgz8z3enj8zb",
    "resource": {
      "type": "k8s_container",
      "labels": {
        "location": "us-central1-a",
        "namespace_name": "default",
        "cluster_name": "snakemake",
        "project_id": "ccgp-ucsc",
        "container_name": "snakejob-6402cff2-af74-5e16-99eb-6b75ef8648e7",
        "pod_name": "snakejob-6402cff2-af74-5e16-99eb-6b75ef8648e7"
      }
    },
    "timestamp": "2022-02-26T21:41:35.828494305Z",
    "severity": "ERROR",
    "labels": {
      "k8s-pod/app": "snakemake",
      "compute.googleapis.com/resource_name": "gke-snakemake-homestar-runner3-b2e8cbe5-wvvh"
    },
    "logName": "projects/ccgp-ucsc/logs/stderr",
    "receiveTimestamp": "2022-02-26T21:41:38.002988901Z"
  },
  {
    "textPayload": "Complete log: /workdir/.snakemake/log/2022-02-26T213417.888994.snakemake.log\n",
    "insertId": "uk8redlmaltekpns",
    "resource": {
      "type": "k8s_container",
      "labels": {
        "pod_name": "snakejob-6402cff2-af74-5e16-99eb-6b75ef8648e7",
        "container_name": "snakejob-6402cff2-af74-5e16-99eb-6b75ef8648e7",
        "cluster_name": "snakemake",
        "project_id": "ccgp-ucsc",
        "location": "us-central1-a",
        "namespace_name": "default"
      }
    },
    "timestamp": "2022-02-26T21:41:35.829045375Z",
    "severity": "ERROR",
    "labels": {
      "compute.googleapis.com/resource_name": "gke-snakemake-homestar-runner3-b2e8cbe5-wvvh",
      "k8s-pod/app": "snakemake"
    },
    "logName": "projects/ccgp-ucsc/logs/stderr",
    "receiveTimestamp": "2022-02-26T21:41:38.002988901Z"
...
  {
    "protoPayload": {
      "@type": "type.googleapis.com/google.cloud.audit.AuditLog",
      "authenticationInfo": {
        "principalEmail": "system:serviceaccount:kube-system:pod-garbage-collector"
      },
      "authorizationInfo": [
        {
          "granted": true,
          "permission": "io.k8s.core.v1.pods.delete",
          "resource": "core/v1/namespaces/default/pods/snakejob-6402cff2-af74-5e16-99eb-6b75ef8648e7"
        }
      ],
      "methodName": "io.k8s.core.v1.pods.delete",
      "request": {
        "@type": "core.k8s.io/v1.DeleteOptions",
        "apiVersion": "v1",
        "gracePeriodSeconds": 0,
        "kind": "DeleteOptions"
      },
      "requestMetadata": {
        "callerIp": "10.128.1.36",
        "callerSuppliedUserAgent": "kube-controller-manager/v1.21.6 (linux/amd64) kubernetes/7ce0f9f/system:serviceaccount:kube-system:pod-garbage-collector"
      },
      "resourceName": "core/v1/namespaces/default/pods/snakejob-6402cff2-af74-5e16-99eb-6b75ef8648e7",
      "response": {
        "@type": "core.k8s.io/v1.Pod",
        "apiVersion": "v1",
        "kind": "Pod",
        "metadata": {
          "creationTimestamp": "2022-02-26T21:32:21Z",
          "deletionGracePeriodSeconds": 0,
          "deletionTimestamp": "2022-02-26T21:57:00Z",
          "labels": {
            "app": "snakemake"
          },
          "managedFields": [
            {
```
###Good pod:
```
{
    "textPayload": "Finished job 0.\n",
    "insertId": "kyevj6phm14w7x29",
    "resource": {
      "type": "k8s_container",
      "labels": {
        "namespace_name": "default",
        "pod_name": "snakejob-374ab5eb-162b-5ed8-8aec-65f6f2c3d0a1",
        "cluster_name": "snakemake",
        "container_name": "snakejob-374ab5eb-162b-5ed8-8aec-65f6f2c3d0a1",
        "location": "us-central1-a",
        "project_id": "ccgp-ucsc"
      }
    },
    "timestamp": "2022-02-26T21:43:36.605458336Z",
    "severity": "ERROR",
    "labels": {
      "compute.googleapis.com/resource_name": "gke-snakemake-homestar-runner3-b2e8cbe5-4d2d",
      "k8s-pod/app": "snakemake"
    },
    "logName": "projects/ccgp-ucsc/logs/stderr",
    "receiveTimestamp": "2022-02-26T21:43:41.258457922Z"
  },
  {
    "textPayload": "1 of 1 steps (100%) done\n",
    "insertId": "7m30qvn69661jizg",
    "resource": {
      "type": "k8s_container",
      "labels": {
        "project_id": "ccgp-ucsc",
        "container_name": "snakejob-374ab5eb-162b-5ed8-8aec-65f6f2c3d0a1",
        "namespace_name": "default",
        "pod_name": "snakejob-374ab5eb-162b-5ed8-8aec-65f6f2c3d0a1",
        "location": "us-central1-a",
        "cluster_name": "snakemake"
      }
    },
    "timestamp": "2022-02-26T21:43:36.605462716Z",
    "severity": "ERROR",
    "labels": {
      "k8s-pod/app": "snakemake",
      "compute.googleapis.com/resource_name": "gke-snakemake-homestar-runner3-b2e8cbe5-4d2d"
    },
    "logName": "projects/ccgp-ucsc/logs/stderr",
    "receiveTimestamp": "2022-02-26T21:43:41.258457922Z"
  },
...
  {
    "protoPayload": {
      "@type": "type.googleapis.com/google.cloud.audit.AuditLog",
      "authenticationInfo": {
        "principalEmail": "cmirchan@ucsc.edu"
      },
      "authorizationInfo": [
        {
          "granted": true,
          "permission": "io.k8s.core.v1.pods.delete",
          "resource": "core/v1/namespaces/default/pods/snakejob-374ab5eb-162b-5ed8-8aec-65f6f2c3d0a1"
        }
      ],
      "methodName": "io.k8s.core.v1.pods.delete",
      "request": {
        "@type": "core.k8s.io/v1.DeleteOptions",
        "apiVersion": "v1",
        "kind": "DeleteOptions"
      },
      "requestMetadata": {
        "callerIp": "***.***.***.***",
        "callerSuppliedUserAgent": "OpenAPI-Generator/22.6.0/python"
      },
      "resourceName": "core/v1/namespaces/default/pods/snakejob-374ab5eb-162b-5ed8-8aec-65f6f2c3d0a1",
      "response": {
        "@type": "core.k8s.io/v1.Pod",
        "apiVersion": "v1",
        "kind": "Pod",
        "metadata": {
          "creationTimestamp": "2022-02-26T21:32:21Z",
          "deletionGracePeriodSeconds": 0,
          "deletionTimestamp": "2022-02-26T21:58:34Z",
          "labels": {
            "app": "snakemake"
          },
          "managedFields": [
            {
  
```

**Comment by johanneskoester** [comment]
The KeyError you have in the initial post is related to a bug that was fixed with Snakemake 7.0, which could lead to duplicate jobs. As Snakemake deletes output files before starting a job, it could be that the duplicate job deleted the output files in the moment the first job was complete, leading to a missing output file exception. Could you try again with Snakemake 7.0?

**Comment by cademirch** [comment]
> The KeyError you have in the initial post is related to a bug that was fixed with Snakemake 7.0, which could lead to duplicate jobs. As Snakemake deletes output files before starting a job, it could be that the duplicate job deleted the output files in the moment the first job was complete, leading to a missing output file exception. Could you try again with Snakemake 7.0?

The KeyError appears to be fixed in 7.0. As for the MissingOutput there is discussion in #1444 about it.

**Comment by johanneskoester** [comment]
Closing in favor of #1460.
